### PR TITLE
finish `RegionKind` renaming

### PR DESCRIPTION
**Pull request by lcnr**
second step of https://github.com/rust-lang/types-team/issues/95

continues the work from #117876. While working on this and I encountered a bunch of further cleanup which I'll either open a tracking issue for or will do in a separate PR:
- [ ] rewrite the `RegionKind` docs, they still talk about `ReEmpty` and are generally out of date
- [x] rename `DescriptionCtx` to `DescriptionCtxt` (already inconsistent)
- [x] what is `CheckRegions::Bound`?
- [ ] `collect_late_bound_regions` et al
- [x] `erase_late_bound_regions` -> `instantiate_bound_regions_with_erased`?
- [x] `EraseEarlyRegions` visitor should be removed, feels duplicate

r? @BoxyUwU

### CODE REVIEW

File: compiler/rustc_infer/src/infer/mod.rs [hunk 1]
```diff
@@ -508,7 +508,10 @@ pub enum RegionVariableOrigin {
     Coercion(Span),
 
     /// Region variables created as the values for early-bound regions.
-    EarlyBoundRegion(Span, Symbol),
+    ///
+    /// FIXME(@lcnr): This can also store a `DefId`, similar to
+    /// `TypeVariableOriginKind::TypeParameterDefinition`.
+    RegionParameterDefinition(Span, Symbol),
```
Comment (BoxyUwU) [inline compiler/rustc_infer/src/infer/mod.rs:514]: is this used for late and early bound parameters?

Comment (lcnr) [reply compiler/rustc_infer/src/infer/mod.rs:514]: only early bound, copied he name from `TypeVariableOriginKind`

**Pull request merged**